File: three-language-version/lessons/09-pattern-matching/pattern_matching.py
"""
Lesson 9: Pattern Matching - Python Examples
Python uses structural pattern matching (match/case) since 3.10
"""

# =============================================================================
# BASIC PATTERN MATCHING (Python 3.10+)
# =============================================================================

def describe_type(value):
    """Match on different types"""
    match value:
        case int():
            return "It's an integer"
        case float():
            return "It's a float"
        case str():
            return "It's a string"
        case list():
            return "It's a list"
        case _:
            return "It's something else"

print("=== Type Matching ===")
print(describe_type(42))
print(describe_type(3.14))
print(describe_type("hello"))
print(describe_type([1, 2, 3]))


# =============================================================================
# MATCHING LITERAL VALUES
# =============================================================================

def http_status(code):
    """Match HTTP status codes"""
    match code:
        case 200:
            return "OK"
        case 201:
            return "Created"
        case 400:
            return "Bad Request"
        case 404:
            return "Not Found"
        case 500:
            return "Internal Server Error"
        case _:
            return f"Unknown status: {code}"

print("\n=== Literal Matching ===")
for code in [200, 404, 418]:
    print(f"HTTP {code}: {http_status(code)}")


# =============================================================================
# DESTRUCTURING SEQUENCES
# =============================================================================

def process_point(point):
    """Match and destructure point coordinates"""
    match point:
        case (0, 0):
            return "Origin"
        case (0, y):
            return f"On Y-axis at y={y}"
        case (x, 0):
            return f"On X-axis at x={x}"
        case (x, y):
            return f"Point at ({x}, {y})"
        case _:
            return "Not a valid point"

print("\n=== Sequence Destructuring ===")
points = [(0, 0), (0, 5), (3, 0), (3, 4)]
for p in points:
    print(f"{p} -> {process_point(p)}")


# =============================================================================
# MATCHING LISTS WITH PATTERNS
# =============================================================================

def analyze_list(lst):
    """Pattern match on list structure"""
    match lst:
        case []:
            return "Empty list"
        case [x]:
            return f"Single element: {x}"
        case [x, y]:
            return f"Two elements: {x} and {y}"
        case [x, y, *rest]:
            return f"Starts with {x}, {y}, then {len(rest)} more"
        case _:
            return "Not a list"

print("\n=== List Pattern Matching ===")
lists = [[], [1], [1, 2], [1, 2, 3, 4, 5]]
for lst in lists:
    print(f"{lst} -> {analyze_list(lst)}")


# =============================================================================
# MATCHING WITH GUARDS
# =============================================================================

def categorize_age(age):
    """Match with conditional guards"""
    match age:
        case n if n < 0:
            return "Invalid age"
        case n if n < 13:
            return "Child"
        case n if n < 20:
            return "Teenager"
        case n if n < 65:
            return "Adult"
        case _:
            return "Senior"

print("\n=== Guards ===")
ages = [-1, 5, 15, 30, 70]
for age in ages:
    print(f"Age {age}: {categorize_age(age)}")


# =============================================================================
# MATCHING DICTIONARIES
# =============================================================================

def process_event(event):
    """Match dictionary structure"""
    match event:
        case {"type": "click", "x": x, "y": y}:
            return f"Click at ({x}, {y})"
        case {"type": "keypress", "key": k}:
            return f"Key pressed: {k}"
        case {"type": "scroll", "direction": d}:
            return f"Scroll {d}"
        case {"type": t}:
            return f"Unknown event type: {t}"
        case _:
            return "Invalid event"

print("\n=== Dictionary Matching ===")
events = [
    {"type": "click", "x": 100, "y": 200},
    {"type": "keypress", "key": "Enter"},
    {"type": "scroll", "direction": "down"},
    {"type": "hover"},
]
for event in events:
    print(f"{event} -> {process_event(event)}")


# =============================================================================
# MATCHING OBJECTS (Classes)
# =============================================================================

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

class Circle:
    def __init__(self, center, radius):
        self.center = center
        self.radius = radius

class Rectangle:
    def __init__(self, corner, width, height):
        self.corner = corner
        self.width = width
        self.height = height

def describe_shape(shape):
    """Match on object types and attributes"""
    match shape:
        case Circle(center=Point(x=0, y=0), radius=r):
            return f"Circle centered at origin with radius {r}"
        case Circle(radius=r):
            return f"Circle with radius {r}"
        case Rectangle(width=w, height=h) if w == h:
            return f"Square with side {w}"
        case Rectangle(width=w, height=h):
            return f"Rectangle {w}x{h}"
        case _:
            return "Unknown shape"

print("\n=== Object Matching ===")
shapes = [
    Circle(Point(0, 0), 5),
    Circle(Point(1, 1), 3),
    Rectangle(Point(0, 0), 4, 4),
    Rectangle(Point(0, 0), 3, 5),
]
for shape in shapes:
    print(f"{type(shape).__name__} -> {describe_shape(shape)}")


# =============================================================================
# ALTERNATIVE: Traditional Approach (Before Python 3.10)
# =============================================================================

def process_point_traditional(point):
    """Traditional approach without match/case"""
    if not isinstance(point, tuple) or len(point) != 2:
        return "Not a valid point"

    x, y = point  # Manual destructuring

    if x == 0 and y == 0:
        return "Origin"
    elif x == 0:
        return f"On Y-axis at y={y}"
    elif y == 0:
        return f"On X-axis at x={x}"
    else:
        return f"Point at ({x}, {y})"

print("\n=== Traditional Approach (pre-3.10) ===")
print("Same functionality, but more verbose")
print(process_point_traditional((0, 0)))
print(process_point_traditional((3, 4)))


# =============================================================================
# RECURSIVE PATTERN MATCHING
# =============================================================================

def evaluate(expr):
    """Evaluate simple expressions using pattern matching"""
    match expr:
        case int(n):
            return n
        case float(n):
            return n
        case ('+', left, right):
            return evaluate(left) + evaluate(right)
        case ('-', left, right):
            return evaluate(left) - evaluate(right)
        case ('*', left, right):
            return evaluate(left) * evaluate(right)
        case ('/', left, right):
            return evaluate(left) / evaluate(right)
        case _:
            raise ValueError(f"Unknown expression: {expr}")

print("\n=== Expression Evaluation ===")
# (3 + 4) * 2
expr = ('*', ('+', 3, 4), 2)
print(f"{expr} = {evaluate(expr)}")

# 10 / (5 - 3)
expr = ('/', 10, ('-', 5, 3))
print(f"{expr} = {evaluate(expr)}")

File: three-language-version/lessons/09-pattern-matching/pattern_matching.cpp
/**
 * Lesson 9: Pattern Matching - C++ Examples
 * C++ uses std::variant and std::visit for type-safe pattern matching
 */

#include <iostream>
#include <variant>
#include <string>
#include <vector>
#include <optional>
#include <functional>

// =============================================================================
// VARIANT-BASED PATTERN MATCHING
// =============================================================================

// Define a variant that can hold different types
using Value = std::variant<int, double, std::string, std::vector<int>>;

std::string describe_value(const Value& v) {
    return std::visit([](auto&& arg) -> std::string {
        using T = std::decay_t<decltype(arg)>;
        if constexpr (std::is_same_v<T, int>) {
            return "Integer: " + std::to_string(arg);
        } else if constexpr (std::is_same_v<T, double>) {
            return "Double: " + std::to_string(arg);
        } else if constexpr (std::is_same_v<T, std::string>) {
            return "String: " + arg;
        } else if constexpr (std::is_same_v<T, std::vector<int>>) {
            return "Vector with " + std::to_string(arg.size()) + " elements";
        }
    }, v);
}

// =============================================================================
// SHAPES EXAMPLE (Classic pattern matching use case)
// =============================================================================

struct Point {
    double x, y;
};

struct Circle {
    Point center;
    double radius;
};

struct Rectangle {
    Point corner;
    double width, height;
};

struct Triangle {
    Point a, b, c;
};

using Shape = std::variant<Circle, Rectangle, Triangle>;

double area(const Shape& shape) {
    return std::visit([](auto&& s) -> double {
        using T = std::decay_t<decltype(s)>;
        if constexpr (std::is_same_v<T, Circle>) {
            return 3.14159 * s.radius * s.radius;
        } else if constexpr (std::is_same_v<T, Rectangle>) {
            return s.width * s.height;
        } else if constexpr (std::is_same_v<T, Triangle>) {
            // Using cross product formula
            return 0.5 * std::abs(
                (s.b.x - s.a.x) * (s.c.y - s.a.y) -
                (s.c.x - s.a.x) * (s.b.y - s.a.y)
            );
        }
    }, shape);
}

std::string describe_shape(const Shape& shape) {
    return std::visit([](auto&& s) -> std::string {
        using T = std::decay_t<decltype(s)>;
        if constexpr (std::is_same_v<T, Circle>) {
            if (s.center.x == 0 && s.center.y == 0) {
                return "Circle at origin with radius " + std::to_string(s.radius);
            }
            return "Circle at (" + std::to_string(s.center.x) + ", " +
                   std::to_string(s.center.y) + ") with radius " +
                   std::to_string(s.radius);
        } else if constexpr (std::is_same_v<T, Rectangle>) {
            if (s.width == s.height) {
                return "Square with side " + std::to_string(s.width);
            }
            return "Rectangle " + std::to_string(s.width) + "x" +
                   std::to_string(s.height);
        } else if constexpr (std::is_same_v<T, Triangle>) {
            return "Triangle";
        }
    }, shape);
}

// =============================================================================
// EXPRESSION EVALUATION
// =============================================================================

struct Literal { double value; };
struct Add { };
struct Sub { };
struct Mul { };
struct Div { };

using Op = std::variant<Add, Sub, Mul, Div>;

struct BinaryExpr;

using Expr = std::variant<
    Literal,
    std::shared_ptr<BinaryExpr>
>;

struct BinaryExpr {
    Expr left;
    Op op;
    Expr right;
};

double evaluate(const Expr& expr);  // Forward declaration

double apply_op(const Op& op, double left, double right) {
    return std::visit([=](auto&& o) -> double {
        using T = std::decay_t<decltype(o)>;
        if constexpr (std::is_same_v<T, Add>) return left + right;
        else if constexpr (std::is_same_v<T, Sub>) return left - right;
        else if constexpr (std::is_same_v<T, Mul>) return left * right;
        else if constexpr (std::is_same_v<T, Div>) return left / right;
    }, op);
}

double evaluate(const Expr& expr) {
    return std::visit([](auto&& e) -> double {
        using T = std::decay_t<decltype(e)>;
        if constexpr (std::is_same_v<T, Literal>) {
            return e.value;
        } else if constexpr (std::is_same_v<T, std::shared_ptr<BinaryExpr>>) {
            double left = evaluate(e->left);
            double right = evaluate(e->right);
            return apply_op(e->op, left, right);
        }
    }, expr);
}

// =============================================================================
// OPTIONAL PATTERN MATCHING
// =============================================================================

std::optional<int> find_in_vector(const std::vector<int>& vec, int target) {
    for (size_t i = 0; i < vec.size(); i++) {
        if (vec[i] == target) {
            return static_cast<int>(i);
        }
    }
    return std::nullopt;
}

void demonstrate_optional() {
    std::vector<int> nums = {10, 20, 30, 40, 50};

    auto result1 = find_in_vector(nums, 30);
    auto result2 = find_in_vector(nums, 99);

    // "Pattern matching" on optional
    if (result1.has_value()) {
        std::cout << "Found 30 at index " << result1.value() << std::endl;
    } else {
        std::cout << "30 not found" << std::endl;
    }

    // Or using value_or
    std::cout << "Index of 99: " << result2.value_or(-1) << std::endl;
}

// =============================================================================
// HELPER: Overload pattern for cleaner visit syntax
// =============================================================================

template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;

void demonstrate_overloaded() {
    std::cout << "\n=== Using Overloaded Pattern ===" << std::endl;

    Value v1 = 42;
    Value v2 = 3.14;
    Value v3 = std::string("hello");

    for (const auto& v : {v1, v2, v3}) {
        std::visit(overloaded {
            [](int i)    { std::cout << "int: " << i << std::endl; },
            [](double d) { std::cout << "double: " << d << std::endl; },
            [](const std::string& s) { std::cout << "string: " << s << std::endl; },
            [](const std::vector<int>&) { std::cout << "vector" << std::endl; }
        }, v);
    }
}

// =============================================================================
// MAIN
// =============================================================================

int main() {
    std::cout << "=== Variant Pattern Matching ===" << std::endl;
    std::vector<Value> values = {42, 3.14, std::string("hello"), std::vector<int>{1, 2, 3}};
    for (const auto& v : values) {
        std::cout << describe_value(v) << std::endl;
    }

    std::cout << "\n=== Shape Pattern Matching ===" << std::endl;
    std::vector<Shape> shapes = {
        Circle{{0, 0}, 5},
        Circle{{1, 2}, 3},
        Rectangle{{0, 0}, 4, 4},
        Rectangle{{0, 0}, 3, 5},
        Triangle{{0, 0}, {4, 0}, {2, 3}}
    };
    for (const auto& s : shapes) {
        std::cout << describe_shape(s) << " - Area: " << area(s) << std::endl;
    }

    std::cout << "\n=== Expression Evaluation ===" << std::endl;
    // Build: (3 + 4) * 2
    auto expr = std::make_shared<BinaryExpr>(BinaryExpr{
        std::make_shared<BinaryExpr>(BinaryExpr{
            Literal{3}, Add{}, Literal{4}
        }),
        Mul{},
        Literal{2}
    });
    std::cout << "(3 + 4) * 2 = " << evaluate(expr) << std::endl;

    std::cout << "\n=== Optional Pattern Matching ===" << std::endl;
    demonstrate_optional();

    demonstrate_overloaded();

    return 0;
}
